pkg/controller/migmigration/backup.go

```diff
@@ -255,3 +256,154 @@ func (t *Task) bounceResticPod() error {
 
 	return nil
 }
+
+// Annotate all resources with PV action data
+func (t *Task) annotateStorageResources() error {
+	// Get client of source cluster
+	client, err := t.getSourceClient()
+	if err != nil {
+		return err
+	}
+	namespaces := t.PlanResources.MigPlan.Spec.Namespaces
+	pvs := t.PlanResources.MigPlan.Spec.PersistentVolumes
+	for _, pv := range pvs.List {
+		// Update PVs with their action
+		resource := corev1.PersistentVolume{}
+		err := client.Get(
+			context.TODO(),
+			types.NamespacedName{
+				Name: pv.Name,
+			},
+			&resource)
+		if err != nil {
+			return err
+		}
+		if resource.Annotations == nil {
+			resource.Annotations = make(map[string]string)
+		}
+		resource.Annotations[pvAnnotationKey] = pv.Action
+		if resource.Labels == nil {
+			resource.Labels = make(map[string]string)
+		}
+		resource.Labels[pvBackupLabelKey] = pvBackupLabelValue
+		client.Update(context.TODO(), &resource)
+	}
+
+	for _, ns := range namespaces {
+		// Find all pods in our target namespaces
+		list := corev1.PodList{}
+		options := k8sclient.InNamespace(ns)
+		err = client.List(context.TODO(), options, &list)
+		if err != nil {
+			return err
+		}
+		// Loop through all pods to find all volume claims
+		for _, pod := range list.Items {
+			resticVolumes := []string{}
+			for _, volume := range pod.Spec.Volumes {
+				claim := volume.VolumeSource.PersistentVolumeClaim
+				if claim == nil {
+					continue
+				}
+				pvc := corev1.PersistentVolumeClaim{}
+				ref := types.NamespacedName{
+					Namespace: ns,
+					Name:      claim.ClaimName,
+				}
+				// Get PVC and update annotation
+				err = client.Get(context.TODO(), ref, &pvc)
+				if pvc.Annotations == nil {
+					pvc.Annotations = make(map[string]string)
+				}
+				pvName := pvc.Spec.VolumeName
+				action := findPVCAction(pvs, pvName)
+				pvc.Annotations[pvAnnotationKey] = action
+				if pvc.Labels == nil {
+					pvc.Labels = make(map[string]string)
+				}
+				pvc.Labels[pvBackupLabelKey] = pvBackupLabelValue
+				err = client.Update(context.TODO(), &pvc)
+				if action == migapi.PvCopyAction {
+					resticVolumes = append(resticVolumes, volume.Name)
+				}
+			}
+			if len(resticVolumes) > 0 {
+				if pod.Annotations == nil {
+					pod.Annotations = make(map[string]string)
+				}
+				pod.Annotations[resticPvBackupAnnotationKey] = strings.Join(resticVolumes[:], ",")
+				if pod.Labels == nil {
+					pod.Labels = make(map[string]string)
+				}
+				pod.Labels[pvBackupLabelKey] = pvBackupLabelValue
+				err = client.Update(context.TODO(), &pod)
+			}
+		}
+	}
+
+	return nil
+}
+
+// Removes temporary annotations and labels used before backing up storage resources
+func (t *Task) removeStorageResourceAnnotations() error {
```

Comment:
I think this method is wiping the storage annotations off every PV, PVC, and Pod cluster-wide. I may be missing where we are filtering on namespaces, but it feels dangerous to combine:
- a non-unique label used for PV action indication
- cluster-wide deletion of said label when one migration is finished, given that there may be other migrations on the cluster which will also be labeling things (and potentially labeling them differently).

I wonder if we need to 
- Be more fine-grained when wiping annotations 
- Have a MigPlan-unique or MigMigration-unique label indicating the action that should be taken on items.

This may not bite us immediately, since it looks like the annotations are applied and then removed only during an active migration. Being more discriminatory when cleaning up seems like it might be a must-have here, I'm not sure if the unique label is needed. Still thinking.

Would like to hear your take @dymurray @pranavgaikwad

Reply:
Good point. I did not give it a thought when I first wrote this. There should be a unique key to each backup task probably the backup id ?

Reply:
There is a unique backup key, but it doesn't yet exist at the time we call `annotateStorageResources`. Adding a label that references the migration would make sense, though. That way `removeStorageResourceAnnotations` could filter on both labels and only remove the annotations that were added for this particular migration.